Added new benchmarks for ReachSafety-arrays, -loops categories.

Dipali-Bhutada opened the pull request:
Added benchmarks manipulating arrays with tiling.

mchalupa commented on c/array-tiling/mbpr2_true-unreach-call.c:
```diff
@@ -0,0 +1,49 @@
+extern void __VERIFIER_assume(int);
+void __VERIFIER_assert(int cond) { if(!(cond)) { ERROR: __VERIFIER_error(); } }
+
+#define CELLCOUNT
```
I guess here you forgot to define the number.

mchalupa replied:
(As well as in other versions of this benchmark)

Dipali-Bhutada replied:
Ok. Thanks. I will upload the updated files.

mchalupa commented on c/array-tiling/mlceu_false-unreach-call.c:
```diff
@@ -0,0 +1,29 @@
+extern void __VERIFIER_assume(int);
+void __VERIFIER_assert(int cond) { if(!(cond)) { ERROR: __VERIFIER_error(); } }
+
+#define SIZE
```
The same for most of the `SIZE` definitions.

Dipali-Bhutada replied:
Ok. Thanks. I will upload the updated files.

mchalupa commented on c/array-tiling/mbpr2_true-unreach-call.c:
```diff
@@ -0,0 +1,49 @@
+
+#define CELLCOUNT
+
+void main()
```
Perhaps this is a nitpick, but the prototype of main should be `int main(void)` in this case. From the C standard:

```
It shall be defined with a return type of int and with no parameters:

int main(void) { /* ... */ }

or with two parameters (referred to here as argc and argv, though any names may be
used, as they are local to the function in which they are declared):

int main(int argc, char *argv[]) { /* ... */ }

or equivalent;
```

Dipali-Bhutada replied:
Ok. Thanks. I will upload the updated files.

PhilippWendler commented:
Please check the following list (documentation [here](https://github.com/sosy-lab/sv-benchmarks/blob/d58d7c0627422bd911319218beae7779b425eeea/CONTRIBUTING.md)):
- [ ] programs added to new and appropriately named directory
- [ ] license present and acceptable (either in separate file or as comment at beginning of program)
- [ ] contributed-by present (either in README file or as comment at beginning of program)

- [ ] programs added to a `.set` file of an existing category, or new sub-category established (if justified)
- [ ] intended property matches the corresponding `.prp` file
- [ ] expected answer in file names according to convention

<!-- For C programs: -->
- [ ] architecture (32 bit vs. 64 bit) matches the corresponding `.cfg` file
- [ ] original sources present
- [ ] preprocessed files present
- [ ] preprocessed files generated with correct architecture
- [ ] Makefile added with correct content and without overly broad suppression of warnings

Dipali-Bhutada commented:
Hey, I have uploaded the updated benchmarks with new pull request (No. 474). 
Shall I upload them here?

PhilippWendler commented:
In the future, please do not create a new PR for fixes and improvements to an existing PR. You can add new commits to an existing PR by pushing to the branch from which you have created the PR.

This helps keeping the discussion in one place.

Dipali-Bhutada commented:
Okay. Apologies for inconvenience.
Shall we continue with PR 474 now? "Reopen and comment" button seems to be disabled here.

PhilippWendler commented:
You can't reopen this PR because #474 is open, and both were created from the same branch.

Dipali-Bhutada commented:
Okay then I will close #474. And upload latest files here. Will that be ok?

PhilippWendler commented:
Sure.

Dipali-Bhutada commented:
Thanks for consideration.

Dipali-Bhutada commented:
All checks have been passed now. Please review and confirm the verification tasks submitted.

Dipali-Bhutada commented:
Hi, Could somebody please review our benchmarks? They are pending for review since 20 days.

peterschrammel commented on c/loop-acceleration/simple_vardep_true-uncreach-call2_true-termination.c:
```diff
@@ -0,0 +1,26 @@
+/* Bechmark used to verify Chimdyalwar, Bharti, et al. "VeriAbs: Verification by abstraction (competition contribution)." 
```
typo `Benchmark`

peterschrammel commented on c/loop-acceleration/simple_vardep_true-uncreach-call1_true-termination.c:
```diff
@@ -0,0 +1,27 @@
+/* Bechmark used to verify Chimdyalwar, Bharti, et al. "VeriAbs: Verification by abstraction (competition contribution)." 
```
typo `Benchmark`

peterschrammel commented on c/loop-acceleration/simple_array_index_value_true-unreach-call4_true-termination.c:
```diff
@@ -0,0 +1,45 @@
+/* Bechmark used to verify Chimdyalwar, Bharti, et al. "VeriAbs: Verification by abstraction (competition contribution)." 
```
typo `Benchmark`

peterschrammel commented on c/loop-acceleration/simple_array_index_value_true-unreach-call3_true-termination.c:
```diff
@@ -0,0 +1,32 @@
+/* Bechmark used to verify Chimdyalwar, Bharti, et al. "VeriAbs: Verification by abstraction (competition contribution)." 
```
typo `Benchmark` (in all .c files)

peterschrammel commented on c/array-tiling/rewnifrev_true-unreach-call.c:
```diff
@@ -0,0 +1,28 @@
+
+		for( i=SIZE-1; i>=0; i-- )
+		{
+			a[i-1] = i-2;
```
Illegal memory access in the last iteration (when i==0).

peterschrammel commented on c/array-tiling/rewnifrev2_true-unreach-call.c:
```diff
@@ -0,0 +1,28 @@
+
+		for( i=SIZE-2; i >= -1; i-- )
+		{
+			a[i] = i;
```
Illegal memory access in the last iteration (when i==-1).

peterschrammel replied:
The problem is still there.

peterschrammel commented on c/array-examples/copysome2_false-unreach-call.c:
```diff
@@ -0,0 +1,37 @@
+      if (i != z)
+         a3[i] = a2[i];
+      else
+          a3[i] = a1[i];
```
More uniform code formatting would be appreciated.

peterschrammel commented on c/array-examples/copysome1_false-unreach-call.i:
```diff
@@ -0,0 +1,20 @@
+extern void __VERIFIER_error() __attribute__ ((__noreturn__));
```
In my opinion, the preprocessed files don't need to be provided here. The .set files can have patterns for .c files.

PhilippWendler replied:
The original file contains a `#define`. So either the preprocessed file needs to be kept, or the `#define` removed.

Dipali-Bhutada replied:
We would like to keep #define. And prepocessed file is already present.

Dipali-Bhutada commented:
Hi @peterschrammel, Thank you for reviewing the benchmarks.  I have addressed all the issues mentioned above. Please review and confirm.

Dipali-Bhutada commented:
It seems that ReachSafety-Arrays-new.set is newly added file. It expects *.yml files in array-example folder. Therefore Travis CI "Sanity Checks" build is failing for our examples. 
I had changed  ReachSafety-Arrays.set to accommodate our benchmarks earlier.  Do I need to change ReachSafety-Arrays-new.set as well?

PhilippWendler commented:
No, the [log](https://travis-ci.org/sosy-lab/sv-benchmarks/jobs/293118858) shows that the script does not complain about `ReachSafety-Arrays-new.set`, but about files in `loop-acceleration` that have the illegal result `_true-uncreach-call2`.

Btw., why are you adding benchmarks to an existing directory? There is a `README` in this directory that claims that these benchmarks are from CBMC. I guess this would not be true for your benchmarks, so they need to go somewhere else. If you are submitting two different kinds of benchmarks, I guess it would be best to have two PRs (the title and description of this PR does not even mention loop-acceleration benchmarks).

Dipali-Bhutada commented:
I misinterpreted, that name of directory suggest the kind of program it should include. Since programs I added were used in experimenting our loop-acceleration technique, I added them to existing directory. 

I have submitted,
7  benchmarks in loop category - all in loop-acceleration folder
32 benchmarks in array category                               
           4 in array-examples folder
           28 in array-tiling folder.

So Now, 28 programs in array-tiling folder are correctly placed. However, 7 in loop-acceleration folder, and 4 in array-examples folder need to be relocated in new folders with relevant names and READMEs.

Should I be raising 3 different pull requests for these 3 folder? Or will it be fine, if I do these changes in the same PR ?

PhilippWendler commented on c/array-programs/Makefile:
```diff
@@ -0,0 +1,6 @@
+LEVEL := ../
+
+CLANG_WARNINGS := \
+	-Wno-error=uninitialized \
```
Is this `no-error=uninitialized` really necessary? It would be better to fix the benchmarks instead of suppressing the warning, because this is probably undefined behavior.

Dipali-Bhutada replied:
Arrays in the programs are to be left uninitialised. Therefore, "no-error=uninitialized" is necessary.

PhilippWendler replied:
Oh, but having undefined behavior is not allowed, and this indeed occurs in these benchmarks.

You need to ensure that all array cells that are read have been initialized before.
Note that in files like the `copysome` programs, this even means you need to initialize at least some of the `a2` cells. Fixing all compiler warnings is not enough, because compilers are not guaranteed to find all undefined behavior.

If you want non-deterministic values, you need to call `__VERIFIER_nondet_int` in a loop.

So please ensure that there is no undefined behavior and remove the compiler warning.

Dipali-Bhutada commented:
Have addressed the issues found.

dbeyer commented:
There is one day left to get this PR fixed. All the check boxes are yet unmarked.

Dipali-Bhutada commented:
All the requirements in checklist are fulfilled. However, I can't edit the checklist to mark the checkbox. 
Please let me know how do I do that?

tautschnig commented:
@Dipali-Bhutada could you please apply some sensible squashing to avoid having to review 75 commits, where later commits partly supersede earlier ones?

PhilippWendler commented:
@tautschnig Usually squashing would be nice, yes, but you can also simply look at the[ combined diff over all commits](https://github.com/sosy-lab/sv-benchmarks/pull/473/files).

tautschnig commented:
> but you can also simply look at the combined diff over all commits.

Of course I could, but could we please have some minimal quality standards when introducing new benchmarks? I've had to go back to the git history way too often to figure out what's going on with certain benchmarks. If people think I'm setting a too high bar then I'd argue that the use of git is inappropriate and we should instead revert to some plain file sharing.

Dipali-Bhutada commented:
I am sorry for the inconvenience caused to you. However, I am handling this PR through Git Web browser, not through command line/ desktop application. Later ones are more convenient for squashing than web browser Git (This is my experience). Even I was frustrated to delete all the files from folder one by one and commit each time. However, I did not find any quick help to delete multiple files at ones / rename a folder through web browser.  (Snapshot refers to the solution that I found and applied).
![git_issue](https://user-images.githubusercontent.com/32162607/32874538-e541f440-cab8-11e7-93ec-d330e480ce12.png)
 
I once again apologies for trouble caused to you while reviewing. Somehow, I had to choose git web GUI over command line.

Dipali-Bhutada commented:
@peterschrammel changes requested by you have been addressed. They are shown to be outdated now. Please tell me if anything is to be done from my end.

peterschrammel commented:
Thanks @Dipali-Bhutada for addressing my comments. 
@dbeyer might be able to squash it on merge.
